components/notebook-controller/Dockerfile

```diff
@@ -3,32 +3,62 @@
 # The Docker context is expected to be:
 #
 # ${PATH_TO_KUBEFLOW/KUBEFLOW repo}/components
-#
-# This is necessary because the Jupyter controller now depends on
-# components/common
+
+
+# Build arguments
+ARG SOURCE_CODE=.
 ARG GOLANG_VERSION=1.20
-FROM golang:${GOLANG_VERSION} as builder
 
+# Use ubi8/go-toolset as base image
+FROM registry.redhat.io/ubi8/go-toolset:${GOLANG_VERSION} as builder
+
+## Build args to be used at this step
+ARG SOURCE_CODE
+
+# Set building workdir
 WORKDIR /workspace
 
 # Copy the Go Modules manifests
-COPY notebook-controller /workspace/notebook-controller
-COPY common /workspace/common
-
-# cache deps before building and copying source so that we don't need to re-download as much
-# and so that source changes don't invalidate our downloaded layer
-RUN cd /workspace/notebook-controller && go mod download all
+COPY ${SOURCE_CODE}/notebook-controller ./notebook-controller
+# This is necessary because the Jupyter controller now depends on
+# components/common
+COPY ${SOURCE_CODE}/common ./common
 
+# Update building workdir
 WORKDIR /workspace/notebook-controller
 
-# Build
-RUN CGO_ENABLED=0 GOOS=linux GO111MODULE=on go build -a -mod=mod -o manager main.go
+## Build the kf-notebook-controller
+USER root
+
+RUN if [ -z ${CACHITO_ENV_FILE} ]; then \
+       go mod download all; \
+    else \
+       source ${CACHITO_ENV_FILE}; \
+    fi
+
+RUN CGO_ENABLED=0 GOOS=linux GO111MODULE=on go build -a -mod=mod \
+        -o ./bin/manager main.go
+
+# Use ubi8/ubi-minimal as base image
+FROM registry.redhat.io/ubi8/ubi-minimal:latest
+
+## Install additional packages
+RUN microdnf install -y shadow-utils &&\
+    microdnf clean all
 
-# Use distroless as minimal base image to package the manager binary
-# Refer to https://github.com/GoogleContainerTools/distroless for more details
-FROM gcr.io/distroless/base:debug
-WORKDIR /
-COPY --from=builder /workspace/notebook-controller/manager .
+## Create a non-root user with UID 1001
+RUN useradd --uid 1001 --create-home --user-group --system rhods
+
+## Set workdir directory to user home
+WORKDIR /home/rhods
+
+## Copy kf-notebook-controller-manager binary from builder stage
+COPY --from=builder /workspace/notebook-controller/bin/manager /manager
 COPY --from=builder /workspace/notebook-controller/third_party/license.txt third_party/license.txt
-COPY --from=builder /go/pkg/mod/github.com/hashicorp third_party/hashicorp
-ENTRYPOINT ["/manager"]
+# in builder image, `go env GOPATH` outputs `/opt/app-root/src/go`
+COPY --from=builder /opt/app-root/src/go/pkg/mod/github.com/hashicorp third_party/hashicorp
```

Comment:
I see that inherently we were copying hashicorp packages. However, it seems that in the downstream Dockerfile, we aren't specifically specifying this particular package. So, my question is: why do we need to specify 'hashicorp'?

Furthermore, I noticed because i am working on upgrade of the  'controller-runtime` to v0.17.0', that this package is no longer present in the 'go.sum', causing the Docker build to fail.

Overall looks good to me

Reply:
> I see that inherently we were copying hashicorp packages. However, it seems that in the downstream Dockerfile, we aren't specifically specifying this particular package. So, my question is: why do we need to specify 'hashicorp'?

No idea. I decided to keep it in case it was useful for something. Reviewing again, it seems that this is done due to license concerns
* https://github.com/kubeflow/testing/issues/539
* https://github.com/kubeflow/kubeflow/issues/4581
* https://github.com/kubeflow/kubeflow/pull/4643

> Furthermore, I noticed because i am working on upgrade of the 'controller-runtime` to v0.17.0', that this package is no longer present in the 'go.sum', causing the Docker build to fail.

When that's no longer part of the build, then there are no licensing requirements, obviously. I think the right step is to remove the hashicorp line from the dockerfile here.

Done that now, thanks.

Reply:
Thanks!

/lgtm